File: Sources/VoxLocalCore/Settings/SettingsStore.swift
import Foundation
import Combine

public enum HotkeyMode: String, CaseIterable, Sendable {
    /// Record while the shortcut is held; stop on release.
    case pressAndHold
    /// First press starts, second press stops.
    case toggle
}

public enum InsertionMode: String, CaseIterable, Sendable {
    /// Try Accessibility insertion, then paste, then clipboard.
    case automatic
    /// Always just place the result on the clipboard.
    case clipboardOnly
}

public enum SpokenLanguage: String, CaseIterable, Sendable {
    case auto
    case russian = "ru"
    case english = "en"

    /// Value passed to whisper-cli `-l`.
    public var whisperCode: String { self == .auto ? "auto" : rawValue }
}

/// All persisted user settings. Backed by `UserDefaults` (injectable for
/// tests). Every property writes through immediately.
public final class SettingsStore: ObservableObject {
    private let defaults: UserDefaults

    private enum Key {
        static let hotkeyKeyCode = "hotkey.keyCode"
        static let hotkeyModifiers = "hotkey.modifiers"
        static let hotkeyMode = "hotkey.mode"
        static let inputDeviceUID = "audio.inputDeviceUID"
        static let whisperModel = "whisper.model"
        static let spokenLanguage = "whisper.language"
        static let whisperThreads = "whisper.threads"
        static let removeArtifacts = "whisper.removeArtifacts"
        static let keepModelWarm = "whisper.keepWarm"
        static let historyEnabled = "history.enabled"
        static let refinementEnabled = "refine.enabled"
        static let ollamaEndpoint = "refine.ollamaEndpoint"
        static let ollamaModel = "refine.ollamaModel"
        static let refinementPreset = "refine.preset"
        static let customInstruction = "refine.customInstruction"
        static let refinementTimeout = "refine.timeout"
        static let insertionMode = "insert.mode"
        static let launchAtLogin = "app.launchAtLogin"
        static let interfaceLanguage = "app.interfaceLanguage"
        static let logLevel = "app.logLevel"
        static let onboardingCompleted = "app.onboardingCompleted"
    }

    public static let defaultOllamaEndpoint = "http://127.0.0.1:11434"
    public static let defaultHotkeyKeyCode: UInt32 = 49 // Space
    public static let defaultHotkeyModifiers: UInt32 = 0x0800 // Carbon optionKey

    public init(defaults: UserDefaults = .standard) {
        self.defaults = defaults

        hotkeyKeyCode = UInt32(defaults.object(forKey: Key.hotkeyKeyCode) as? Int ?? Int(Self.defaultHotkeyKeyCode))
        hotkeyModifiers = UInt32(defaults.object(forKey: Key.hotkeyModifiers) as? Int ?? Int(Self.defaultHotkeyModifiers))
        hotkeyMode = HotkeyMode(rawValue: defaults.string(forKey: Key.hotkeyMode) ?? "") ?? .pressAndHold
        inputDeviceUID = defaults.string(forKey: Key.inputDeviceUID)
        whisperModel = defaults.string(forKey: Key.whisperModel) ?? "base"
        spokenLanguage = SpokenLanguage(rawValue: defaults.string(forKey: Key.spokenLanguage) ?? "") ?? .auto
        whisperThreads = defaults.object(forKey: Key.whisperThreads) as? Int
            ?? min(8, max(2, ProcessInfo.processInfo.activeProcessorCount / 2))
        removeArtifacts = defaults.object(forKey: Key.removeArtifacts) as? Bool ?? true
        keepModelWarm = defaults.object(forKey: Key.keepModelWarm) as? Bool ?? false
        historyEnabled = defaults.object(forKey: Key.historyEnabled) as? Bool ?? false
        refinementEnabled = defaults.object(forKey: Key.refinementEnabled) as? Bool ?? false
        ollamaEndpoint = defaults.string(forKey: Key.ollamaEndpoint) ?? Self.defaultOllamaEndpoint
        ollamaModel = defaults.string(forKey: Key.ollamaModel) ?? ""
        refinementPreset = RefinementPreset(rawValue: defaults.string(forKey: Key.refinementPreset) ?? "") ?? .cleanDictation
        customInstruction = defaults.string(forKey: Key.customInstruction) ?? ""
        refinementTimeout = defaults.object(forKey: Key.refinementTimeout) as? Double ?? 20.0
        insertionMode = InsertionMode(rawValue: defaults.string(forKey: Key.insertionMode) ?? "") ?? .automatic
        launchAtLogin = defaults.object(forKey: Key.launchAtLogin) as? Bool ?? false
        interfaceLanguage = L10n.Language(rawValue: defaults.string(forKey: Key.interfaceLanguage) ?? "") ?? .russian
        logLevel = LogLevel(rawValue: defaults.object(forKey: Key.logLevel) as? Int ?? LogLevel.info.rawValue) ?? .info
        onboardingCompleted = defaults.object(forKey: Key.onboardingCompleted) as? Bool ?? false
    }

    @Published public var hotkeyKeyCode: UInt32 {
        didSet { defaults.set(Int(hotkeyKeyCode), forKey: Key.hotkeyKeyCode) }
    }
    @Published public var hotkeyModifiers: UInt32 {
        didSet { defaults.set(Int(hotkeyModifiers), forKey: Key.hotkeyModifiers) }
    }
    @Published public var hotkeyMode: HotkeyMode {
        didSet { defaults.set(hotkeyMode.rawValue, forKey: Key.hotkeyMode) }
    }
    @Published public var inputDeviceUID: String? {
        didSet { defaults.set(inputDeviceUID, forKey: Key.inputDeviceUID) }
    }
    @Published public var whisperModel: String {
        didSet { defaults.set(whisperModel, forKey: Key.whisperModel) }
    }
    @Published public var spokenLanguage: SpokenLanguage {
        didSet { defaults.set(spokenLanguage.rawValue, forKey: Key.spokenLanguage) }
    }
    @Published public var whisperThreads: Int {
        didSet { defaults.set(whisperThreads, forKey: Key.whisperThreads) }
    }
    @Published public var removeArtifacts: Bool {
        didSet { defaults.set(removeArtifacts, forKey: Key.removeArtifacts) }
    }
    @Published public var keepModelWarm: Bool {
        didSet { defaults.set(keepModelWarm, forKey: Key.keepModelWarm) }
    }
    @Published public var historyEnabled: Bool {
        didSet { defaults.set(historyEnabled, forKey: Key.historyEnabled) }
    }
    @Published public var refinementEnabled: Bool {
        didSet { defaults.set(refinementEnabled, forKey: Key.refinementEnabled) }
    }
    @Published public var ollamaEndpoint: String {
        didSet { defaults.set(ollamaEndpoint, forKey: Key.ollamaEndpoint) }
    }
    @Published public var ollamaModel: String {
        didSet { defaults.set(ollamaModel, forKey: Key.ollamaModel) }
    }
    @Published public var refinementPreset: RefinementPreset {
        didSet { defaults.set(refinementPreset.rawValue, forKey: Key.refinementPreset) }
    }
    @Published public var customInstruction: String {
        didSet { defaults.set(customInstruction, forKey: Key.customInstruction) }
    }
    @Published public var refinementTimeout: Double {
        didSet { defaults.set(refinementTimeout, forKey: Key.refinementTimeout) }
    }
    @Published public var insertionMode: InsertionMode {
        didSet { defaults.set(insertionMode.rawValue, forKey: Key.insertionMode) }
    }
    @Published public var launchAtLogin: Bool {
        didSet { defaults.set(launchAtLogin, forKey: Key.launchAtLogin) }
    }
    @Published public var interfaceLanguage: L10n.Language {
        didSet {
            defaults.set(interfaceLanguage.rawValue, forKey: Key.interfaceLanguage)
            L10n.setLanguage(interfaceLanguage)
        }
    }
    @Published public var logLevel: LogLevel {
        didSet {
            defaults.set(logLevel.rawValue, forKey: Key.logLevel)
            Log.shared.level = logLevel
        }
    }
    @Published public var onboardingCompleted: Bool {
        didSet { defaults.set(onboardingCompleted, forKey: Key.onboardingCompleted) }
    }

    public func resetOnboarding() {
        onboardingCompleted = false
    }
}

File: windows/src/VoxLocal.App/Core/Settings/SettingsStore.cs
using System.ComponentModel;
using System.Runtime.CompilerServices;
using System.Text.Json;
using System.Text.Json.Serialization;
using VoxLocal.Core.Hotkeys;
using VoxLocal.Core.Insertion;
using VoxLocal.Core.Refinement;
using VoxLocal.Core.Utilities;

namespace VoxLocal.Core.Settings;

public enum HotkeyMode
{
    /// <summary>Record while the shortcut is held; stop on release.</summary>
    PressAndHold,
    /// <summary>First press starts, second press stops.</summary>
    Toggle,
}

public enum SpokenLanguage
{
    Auto,
    Russian,
    English,
}

public static class SpokenLanguageExtensions
{
    /// <summary>Value passed to whisper-cli -l.</summary>
    public static string WhisperCode(this SpokenLanguage language) => language switch
    {
        SpokenLanguage.Russian => "ru",
        SpokenLanguage.English => "en",
        _ => "auto",
    };
}

/// <summary>
/// All persisted user settings. UserDefaults on macOS → a JSON file at
/// %APPDATA%\VoxLocal\settings.json on Windows (path injectable for tests).
/// Every property change writes through immediately, like didSet did.
/// </summary>
public sealed class SettingsStore : INotifyPropertyChanged
{
    public const string DefaultOllamaEndpoint = "http://127.0.0.1:11434";

    /// <summary>Serialized shape of the settings file.</summary>
    private sealed class Data
    {
        public uint HotkeyKeyCode { get; set; } = KeyCombo.Default.KeyCode;
        public uint HotkeyModifiers { get; set; } = KeyCombo.Default.Modifiers;
        public HotkeyMode HotkeyMode { get; set; } = HotkeyMode.PressAndHold;
        public string? InputDeviceId { get; set; }
        public string WhisperModel { get; set; } = "base";
        public SpokenLanguage SpokenLanguage { get; set; } = SpokenLanguage.Auto;
        public int WhisperThreads { get; set; } = Math.Min(8, Math.Max(2, Environment.ProcessorCount / 2));
        public bool RemoveArtifacts { get; set; } = true;
        public bool RefinementEnabled { get; set; }
        public string OllamaEndpoint { get; set; } = DefaultOllamaEndpoint;
        public string OllamaModel { get; set; } = "";
        public RefinementPreset RefinementPreset { get; set; } = RefinementPreset.CleanDictation;
        public string CustomInstruction { get; set; } = "";
        public double RefinementTimeout { get; set; } = 20.0;
        public InsertionMode InsertionMode { get; set; } = InsertionMode.Auto;
        public bool LaunchAtLogin { get; set; }
        public L10n.Language InterfaceLanguage { get; set; } = L10n.Language.Russian;
        public LogLevel LogLevel { get; set; } = LogLevel.Info;
        public bool OnboardingCompleted { get; set; }
    }

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        WriteIndented = true,
        Converters = { new JsonStringEnumConverter(JsonNamingPolicy.CamelCase) },
    };

    private readonly string _path;
    private readonly Data _data;

    public event PropertyChangedEventHandler? PropertyChanged;

    public SettingsStore(string? path = null)
    {
        _path = path ?? System.IO.Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
            "VoxLocal", "settings.json");
        _data = Load(_path);
        // Re-apply side effects for values loaded from disk.
        L10n.SetLanguage(_data.InterfaceLanguage);
        Log.Shared.Level = _data.LogLevel;
    }

    private static Data Load(string path)
    {
        try
        {
            if (File.Exists(path))
            {
                return JsonSerializer.Deserialize<Data>(File.ReadAllText(path), JsonOptions) ?? new Data();
            }
        }
        catch (Exception e) when (e is IOException or JsonException)
        {
            Log.Shared.Error($"settings load failed, using defaults: {e.Message}");
        }
        return new Data();
    }

    /// <summary>Write-through persistence: atomic temp-file + move.</summary>
    private void Save()
    {
        try
        {
            Directory.CreateDirectory(System.IO.Path.GetDirectoryName(_path)!);
            var tmp = _path + ".tmp";
            File.WriteAllText(tmp, JsonSerializer.Serialize(_data, JsonOptions));
            File.Move(tmp, _path, overwrite: true);
        }
        catch (IOException e)
        {
            Log.Shared.Error($"settings save failed: {e.Message}");
        }
    }

    private void SaveAndNotify([CallerMemberName] string? name = null)
    {
        Save();
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
    }

    public uint HotkeyKeyCode
    {
        get => _data.HotkeyKeyCode;
        set { _data.HotkeyKeyCode = value; SaveAndNotify(); }
    }

    public uint HotkeyModifiers
    {
        get => _data.HotkeyModifiers;
        set { _data.HotkeyModifiers = value; SaveAndNotify(); }
    }

    /// <summary>Convenience accessor pairing key code + modifiers.</summary>
    public KeyCombo Hotkey
    {
        get => new(HotkeyKeyCode, HotkeyModifiers);
        set
        {
            _data.HotkeyKeyCode = value.KeyCode;
            _data.HotkeyModifiers = value.Modifiers;
            SaveAndNotify();
        }
    }

    public HotkeyMode HotkeyMode
    {
        get => _data.HotkeyMode;
        set { _data.HotkeyMode = value; SaveAndNotify(); }
    }

    /// <summary>MMDevice ID of the selected input device (null = system default).</summary>
    public string? InputDeviceId
    {
        get => _data.InputDeviceId;
        set { _data.InputDeviceId = value; SaveAndNotify(); }
    }

    public string WhisperModel
    {
        get => _data.WhisperModel;
        set { _data.WhisperModel = value; SaveAndNotify(); }
    }

    public SpokenLanguage SpokenLanguage
    {
        get => _data.SpokenLanguage;
        set { _data.SpokenLanguage = value; SaveAndNotify(); }
    }

    public int WhisperThreads
    {
        get => _data.WhisperThreads;
        set { _data.WhisperThreads = value; SaveAndNotify(); }
    }

    public bool RemoveArtifacts
    {
        get => _data.RemoveArtifacts;
        set { _data.RemoveArtifacts = value; SaveAndNotify(); }
    }

    public bool RefinementEnabled
    {
        get => _data.RefinementEnabled;
        set { _data.RefinementEnabled = value; SaveAndNotify(); }
    }

    public string OllamaEndpoint
    {
        get => _data.OllamaEndpoint;
        set { _data.OllamaEndpoint = value; SaveAndNotify(); }
    }

    public string OllamaModel
    {
        get => _data.OllamaModel;
        set { _data.OllamaModel = value; SaveAndNotify(); }
    }

    public RefinementPreset RefinementPreset
    {
        get => _data.RefinementPreset;
        set { _data.RefinementPreset = value; SaveAndNotify(); }
    }

    public string CustomInstruction
    {
        get => _data.CustomInstruction;
        set { _data.CustomInstruction = value; SaveAndNotify(); }
    }

    public double RefinementTimeout
    {
        get => _data.RefinementTimeout;
        set { _data.RefinementTimeout = value; SaveAndNotify(); }
    }

    public InsertionMode InsertionMode
    {
        get => _data.InsertionMode;
        set { _data.InsertionMode = value; SaveAndNotify(); }
    }

    /// <summary>Persisted flag; applying it (HKCU Run key) happens in the App module.</summary>
    public bool LaunchAtLogin
    {
        get => _data.LaunchAtLogin;
        set { _data.LaunchAtLogin = value; SaveAndNotify(); }
    }

    public L10n.Language InterfaceLanguage
    {
        get => _data.InterfaceLanguage;
        set
        {
            _data.InterfaceLanguage = value;
            L10n.SetLanguage(value);
            SaveAndNotify();
        }
    }

    public LogLevel LogLevel
    {
        get => _data.LogLevel;
        set
        {
            _data.LogLevel = value;
            Log.Shared.Level = value;
            SaveAndNotify();
        }
    }

    public bool OnboardingCompleted
    {
        get => _data.OnboardingCompleted;
        set { _data.OnboardingCompleted = value; SaveAndNotify(); }
    }

    public void ResetOnboarding() => OnboardingCompleted = false;
}
